lightwallet-client-lib/src/main/java/co/electriccoin/lightwallet/client/internal/LightWalletClientImpl.kt

```diff
@@ -273,18 +277,34 @@ internal class LightWalletClientImpl(
         channelFlow {
             launch {
                 while (true) {
-                    requireChannel()
-                        .createStub(streamingRequestTimeout)
-                        .getMempoolStream(Service.Empty.newBuilder().build())
-                        .map {
-                            Response.Success(RawTransactionUnsafe.new(it))
-                        }.catch {
-                            // do nothing
-                        }.collect {
-                            send(it)
+                    flow {
+                        emitAll(
+                            requireChannel()
+                                .createStub(streamingRequestTimeout)
+                                .getMempoolStream(Service.Empty.newBuilder().build())
+                                .map {
+                                    Response.Success(RawTransactionUnsafe.new(it))
+                                }
+                        )
+                    }.retryWhen { cause, attempt ->
+                        // handle EOF
+                        if (cause is StatusException && cause.status.code == Status.Code.INTERNAL) {
+                            delay(1.seconds)
+                            true
+                        } else {
+                            false
+                        }
+                    }.retryWhen { cause, attempt ->
+                        // handle other exceptions with backoff
+                        if (attempt <= 1) {
+                            delay(1.seconds)
+                        } else {
+                            delay(30.seconds)
                         }
+                        true
+                    }.collect { send(it) }
 
-                    delay(30.seconds)
+                    delay(1.seconds) // collection termination on empty mempool
```

Comment:
Nit: it's not "empty mempool" per se, it's "chain tip changed" (which might result in an empty mempool if the entire mempool contents were included in a newly-mined block, but is not necessarily the case).